src/words/ast/INodeListenerTemp.java

```diff
@@ -13,16 +13,25 @@ public ASTValue eval(WordsEnvironment environment) throws WordsRuntimeException
 		AST predicate = children.get(0);
 		AST statementList = children.get(1);
 		
-		ASTValue predicateValue = predicate.eval(environment);
-
-		// currently only restricted to boolean predicate
-		assert predicateValue.type == ASTValue.ValueType.BOOLEAN : "Predicate has type " + predicateValue.type.toString();
-
-		// temporary listener is created only if the predicate is true
-		if (predicateValue.booleanValue == true) {
-			environment.createListener(predicate, statementList, true);
+		if (predicate instanceof INodeBasicActionPredicate) {
+			ASTValue predicateValue = predicate.eval(environment, statementList);
+			
+			if (predicateValue.booleanValue == true) {
+				environment.createListener(predicate, statementList, true);
+			}
+			return null;
+		} else {
+			ASTValue predicateValue = predicate.eval(environment);
+	
+			// currently only restricted to boolean predicate
+			assert predicateValue.type == ASTValue.ValueType.BOOLEAN : "Predicate has type " + predicateValue.type.toString();
+	
+			// temporary listener is created only if the predicate is true
+			if (predicateValue.booleanValue == true) {
+				environment.createListener(predicate, statementList, true);
+			}
+			
+			return null;
 		}
-		
-		return null;
 	}
 }
```

Comment:
It looks like these edits have only been made to INodeListenerTemp, but not INodeListenerPerm?  I think the two are identical except for the final boolean passed to evironment.createListener(), so perhaps they should be consolidated into one.